### PR TITLE
Adds dns/perf framework for running performance DNS benchmarks

### PR DESCRIPTION
**Pull request by bowei**
Adds a performance testing framework for Kubernetes DNS.

- Python performance runner
- Build for dnsperf docker image

### CODE REVIEW

**Comment by k8s-oncall** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/perf-tests/15)

**Comment by bowei** [comment]
@thockin @MrHohn

File: .travis.yml [hunk 1]
```diff
@@ -22,4 +22,5 @@ script:
     - ./verify/verify-boilerplate.sh
     - ./verify/verify-gofmt.sh
     - ./verify/verify-golint.sh
+    - go get ./...
```
Comment (thockin) [inline .travis.yml]: what does this do?

File: dns/perf/image/Makefile [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		/bin/sh -c "cd /src && ./configure && make -j"
+
+build/image.stamp: Dockerfile.image build/src/dnsperf
+	docker build -t $(IMAGE) -f Dockerfile.image .
+	touch $@
+
+image: build/image.stamp
+
+push:
+	gcloud docker -- push $(IMAGE)
```
Comment (jeremyeder) [inline dns/perf/image/Makefile]: Is this ever used?  Strongly prefer not to tie tests to any particular cloud provider.

File: dns/perf/image/Makefile [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+ARCH ?= amd64
+TAG ?= 1.0
+REGISTRY ?= gcr.io/google_containers
+
+BUILD_IMAGE := kube-dns-perf-client-build
+IMAGE := $(REGISTRY)/kube-dns-perf-client-$(ARCH):$(TAG)
+DNSPERF_URL := ftp://ftp.nominum.com/pub/nominum/dnsperf/2.1.0.0/dnsperf-src-2.1.0.0-1.tar.gz
```
Comment (jeremyeder) [inline dns/perf/image/Makefile]: What is the license for this benchmark?

File: dns/perf/image/histogram.patch [hunk 4]
```diff
@@ -0,0 +1,76 @@
+commit 7156287bd0894e52d02fb1485a95f974642c4beb
```
Comment (jeremyeder) [inline dns/perf/image/histogram.patch]: Is there any chance that nominum would take this patch?

File: dns/perf/params/default.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Default parameters that explores performance space given limited CPU.
```
Comment (jeremyeder) [inline dns/perf/params/default.yaml]: I guess we will need to revisit resource limits for this pod at some point; my thought is that DNS is worthy of not being resource constrained.

**Comment by jeremyeder** [comment]
@sjug

File: dns/perf/params/default.yaml (hunk 5 above)
Comment (bowei) [inline dns/perf/params/default.yaml]: We do not set resource limits -- however, settings limits lets us simulate cases where other load pushes the DNS server to be constrained to its resource request. We want to know under these stress conditions what behavior we can expect from the system.

File: dns/perf/image/histogram.patch (hunk 4 above)
Comment (bowei) [inline dns/perf/image/histogram.patch]: I will contact them.

File: dns/perf/image/Makefile (hunk 2 above)
Comment (bowei) [inline dns/perf/image/Makefile]: I can parameterize the `gcloud` command. However, the image is still hosted on google_containers...

**Comment by thockin** [comment]
`go get` fetches from a remote repo and builds.  Since we already have the
code, maybe you need `go install` ?

File: dns/perf/image/Makefile (hunk 3 above)
Comment (bowei) [inline dns/perf/image/Makefile]: It looks like a BSD-1 license:

```
/*
 * Copyright (C) 2000, 2001  Nominum, Inc.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL
 * INTERNET SOFTWARE CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING
 * FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
 * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
 * WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
 * Copyright (C) 2004 - 2015 Nominum, Inc.
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose with or without fee is hereby granted,
 * provided that the above copyright notice and this permission notice
 * appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND NOMINUM DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL NOMINUM BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
 * OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
```

**Comment by bowei** [comment]
@thockin -- that is the problem unless I am missing something, the network test code depends on the kube go-client, which is not part of the repo?

File: dns/perf/README.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+directory that was created.
+
+### Benchmarking the cluster DNS
+
+You can benchmark the existing cluster DNS by specifying the `--use-cluster-dns`
+flag. (As opposed to the server referenced by `--deployment-yaml`). Note: you
+should be aware that some noise may be introduced by the client running on the
+same pod as a DNS server.
+
+Note: some parameters are skipped if running with the cluster DNS as they do not
```
Comment (MrHohn) [inline dns/perf/README.md]: Should we mention which params will be skipped, or instead say which params will be run?

File: dns/perf/README.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+Use the `ingest` script to parse the results of the runs into a sqlite3
+database.
+
+```sh
+$ ./ingest --db out/db out/latest/*.out
+```
+
+The resulting metrics can then be queried using sqlite3. The schema of the
+database can be shown using `sqlite3 out/db ".schema"`. To run sql queries,
+you can use `sqlite3 out/db < my-query.sql`.
```
Comment (MrHohn) [inline dns/perf/README.md]: Would a sample DB schema help? For giving the users an idea of what this perf test may provide them.

File: dns/perf/README.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+This directory contains scripts used to run a dns performance test in a
+Kubernetes cluster. The performance script `run` benchmarks the performance of a
+single DNS server instance with a synthetic query workload.
+
+# Quickstart
+
+## Running a performance test
+
+Assuming you have a working `kubectl` command:
```
Comment (bowei) [inline dns/perf/README.md]: done

File: dns/perf/README.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  AND results.run_subid = runs.run_subid
+ORDER BY
+  qps ASC;
+```
+
+Additional sql queries can be found in `sql/`.
+
+# Monitoring
+
+Kubernetes kube-dns v1.5+ now exports [Prometheus](http://prometheus.io) metrics
```
Comment (bowei) [inline dns/perf/README.md]: referenced image

File: dns/perf/README.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+max_qps: [500, 1000, 2000, 3000, null]
+# File to take queries from. This is in dnsperf format.
+query_file: ["nx-domain.txt", "outside.txt", "pod-ip.txt", "service.txt"]
+```
+
+# Extending the perf test
+
+## Using a different DNS server
+
+You can give different DNS server yaml to the runner via the `--deployment-yaml`
```
Comment (bowei) [inline dns/perf/README.md]: added

File: dns/perf/cluster/dnsperf.yaml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+# limitations under the License.
+apiVersion: v1
+kind: Pod
+metadata:
+  name: kube-dns-perf-client
+spec:
+  terminationGracePeriodSeconds: 1
+  containers:
+  - command: ["sh", "-c", "while true; do echo `date`: MARK; sleep 10; done"]
+    image: gcr.io/bowei-gke-dev/dnsperf:1.0
```
Comment (bowei) [inline dns/perf/cluster/dnsperf.yaml]: fixed

**Pull request merged**